File: LeetCode/python/290.py
class Solution(object):
    def wordPattern(self, pattern, s):
        """
        :type pattern: str
        :type s: str
        :rtype: bool
        """
        match, rev = {}, {}
        s = s.split(' ')
        if len(s) != len(pattern):
            return False
        for i in range(len(s)):
            if (pattern[i] in match and match[pattern[i]] != s[i]) or (s[i] in rev and rev[s[i]] != pattern[i]):
                return False
            match[pattern[i]] = s[i]
            rev[s[i]] = pattern[i]
        return True
            

File: LeetCode/cpp/290.cpp
class Solution {
public:
    bool wordPattern(string pattern, string str) {
        unordered_map<char, string> seq;
        unordered_map<string, char> rev;
        vector<string> s;
        int b = 0;
        for (int i = 0; i < str.size(); ++i) {
            if (str[i] == ' ') {
                s.push_back(str.substr(b, i - b));
                b = i + 1;
            }
        }
        s.push_back(str.substr(b, str.size() - b));
        if (s.size() != pattern.size())
            return false;
        for (int i = 0; i < pattern.size(); ++i) {
            if ((seq.find(pattern[i]) != seq.end() && seq[pattern[i]] != s[i]) ||
                (rev.find(s[i]) != rev.end() && rev[s[i]] != pattern[i]))
                return false;
            cout << 1;
            seq[pattern[i]] = s[i];
            cout << 2;
            rev[s[i]] = pattern[i];
            cout << 3;
        }
        return true;
    }
};
